contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,145 @@
+
+## Goals
+
+* Allow a CSI volume plugin (that opts-in to new mechanism) to configure how Kubernetes should interact with it.
+* Improve discoverability by users/cluster admins of a CSI volume plugin (that opts-in to new mechanism) deployed on the Kubernetes cluster.
+
+## Non-Goals
+
+* Define how CSI volume plugin should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume plugins deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource will be automatically be installed on Kubernetes clusters.
+
+Upon deployment a plugin must create a new custom resource object.
+
+## Design Details
+
+### `CSIPlugin` Object
+
+#### Proposed API
+
+```go
+// CSIPlugin captures information about a Container Storage Interface (CSI)
+// volume plugin deployed on the cluster.
+//
+// CSIPlugin objects are non-namespaced.
+type CSIPlugin struct {
+    metav1.TypeMeta   `json:",inline"`
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
+
+    // Specification describing the CSI volume plugin and any custom
+    // configuration for it.
+	Spec CSIPluginSpec `json:"spec"`
+
+	// Status of the CSI volume plugin.
+	Status CSIPluginStatus `json:"status,omitempty"`
+}
+
+// CSIPluginList is a collection of CSIPlugin objects.
+type CSIPluginList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of CSIPlugins
+	Items []CSIPlugin `json:"items"`
+}
+
+// CSIPluginSpec is the specification of a CSIPlugin.
+type CSIPluginSpec struct {
+    // Driver indicates the name of the CSI Plugin that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume plugin does not require an attach operation
+    // (because it does not implement the ControllerPublishVolume() method),
+    // and that Kubernetes should not call or wait for any attach operation and
+    // just skip to mounting.
+	SkipAttach bool `json:"skipAttach"`
+}
+
+// CSIPluginStatus is the status of a CSIPlugin.
+type CSIPluginStatus struct {
+    // Indicates the volume plugin has been successfully deployed on the cluster
+    // and is ready to use.
+	Ready bool `json:"ready"`
+}
+
+```
+
+#### CRD Installation
+
+The `CSIPlugin` object schema will checked in to kubernetes/kubernetes repo under a storage directory.
+The `CSIPlugin` `CustomResourceDefinition` (CRD) will be installed by the exiting Kubernetes attach/detach controller that is part of the kube-controller-manager binary.
```

Comment:
I agree. But it would be easier to use an existing controller for the alpha version.

I changed it to say new controller -- if we have enough time, I'll do new controller, otherwise I'll just do existing controller and we can add in a standalone controller later for beta.